rfcs/2020-07-21-2744-vicscript.md

```diff
@@ -0,0 +1,195 @@
+
+Joining events are within scope, but aggregation of those events is not. For example, it would be possible to aggregate events using a transaction transform (or wasm, etc) in a form where vicscript can be used to express exactly how the fields of those events should be joined. However, vicscript itself will not have the capability to temporarily store events and has no awareness of delivery guarantees or transactions.
+
+## Guide Level Proposal
+
+The `vicscript` transform allows you to construct a new event by defining a sequence of mapping statements from a source event. Each mapping statement describes a read-only query operation on the right hand side, and on the left hand side a destination for the query result to be mapped within the resulting event:
+
+```toml
+[transforms.mapper]
+  inputs = [ "foo" ]
+  type = "vicscript"
+  mapping = """
+foo = "hello"
+bar = bar + 10
+baz_is_large = baz > 10
+"""
+```
+
+> When executing a Vicscript mapping the source event is never mutated.
+
+The most common query type is a dot separated path, describing how to reach a target field within the input event:
+
+```coffee
+foo = bar.baz.0.buz
+```
+
+Path sections that include whitespace or other special characters (including `.`) can be quoted:
+
+```coffee
+foo = bar."baz.buz".bev
+
+# Use slash escapes within quotes
+bar = buz."fub\"\\fob".fab
+```
+
+And supports coalescing fields within a path with a pipe operator `|`, where if the first value is null or missing then the next value is selected, and so on:
+
+```coffee
+foo = bar.(baz | buz).bev
+```
+
+You can assign a value to the root of your event by writing to the `root` keyword:
+
+```coffee
+root = nested.object
+```
+
+Vicscript supports float, int, boolean, string, null, array and object literals:
+
+```coffee
+first = 7
+second = false
+third = "string"
+fourth = null
+```
+
+As well as a timestamp type:
+
+```coffee
+# NOTE: The full API for the timestamp type is out of the scope of this RFC. The
+# following example is for demonstrative purposes.
+
+created_at = now().unix()
+# Sets `created_at` to an integer representing the current unix timestamp
+# (in seconds).
+```
+
+Boolean operators and arithmetic galore:
+
+```coffee
+is_big = number > 100
+multiplied = number * 7
+```
+
+Perform assignments conditionally with an `if` statement:
+
+```coffee
+id = id
+sorted_foo = if foo.type() == "array" { foo.sort() }
+```
+
+```text
+In:  {"id":"first","foo":"not an array"}
+Out: {"id":"first"}'
+
+In:  {"id":"second","foo":["c","a","d","b"]}
+Out: {"id":"second","sorted_foo":["a","b","c","d"]}
+```
+
+The resulting event will only contain fields that are explicitly mapped, in order to begin your map with a full copy of the input event structure you can start by assigning `this` to `root`:
```

Comment:
I don't love this pattern. Again, the jq approach here is interesting and doesn't require these special names. A full copy is the default with simple assignment like `.foo = "bar"`, while `{foo: "bar"}` gives you a fresh object. 

I'm also curious if there would be performance implications to this for simple cases like `add_fields`. With enough work we could definitely optimize away the copying, but I'm not sure we really want to sign ourselves up for that.